File: 735. Asteroid Collision/Solution.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack, res = [], []
        for a in asteroids:
            if a < 0:
                while stack and stack[-1] < -a: stack.pop()
                if stack and stack[-1] == -a:
                    stack.pop()
                elif not stack:
                    res.append(a)
            else:
                stack.append(a)
        return res + stack

File: 735. Asteroid Collision/Solution.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>(), res = new ArrayDeque<>();
        for (int a : asteroids) {
            if (a < 0) {
                while (!stack.isEmpty() && stack.peekLast() < -a) {
                    stack.pollLast();
                } 
                if (!stack.isEmpty() && stack.peekLast() == -a) {
                    stack.pollLast();
                } else if (stack.isEmpty()) {
                    res.offerLast(a);
                }
            } else {
                stack.offerLast(a);
            }
        }
        int l = stack.size() + res.size(), i = 0;
        int[] ans = new int[l];
        while (!res.isEmpty()) {
            ans[i] = res.pollFirst();
            i++;
        }
        while (!stack.isEmpty()) {
            ans[i] = stack.pollFirst();
            i++;
        }
        return ans;
    }
}
